File: src/main/java/cspom/compiler/patterns/RemoveAnd.java
package cspom.compiler.patterns;

import java.util.Deque;

import scala.collection.JavaConversions;

import cspom.CSPOM;
import cspom.constraint.CSPOMConstraint;
import cspom.constraint.GeneralConstraint;
import cspom.variable.CSPOMVariable;
import cspom.variable.TrueDomain$;

public final class RemoveAnd implements ConstraintCompiler {

    private final CSPOM problem;
    private final Deque<CSPOMConstraint> constraints;

    public RemoveAnd(final CSPOM problem,
            final Deque<CSPOMConstraint> constraints) {
        this.problem = problem;
        this.constraints = constraints;
    }

    @Override
    public void compile(final CSPOMConstraint constraint) {
        if ("and".equals(constraint.description())
                && constraint instanceof GeneralConstraint) {
            for (CSPOMVariable<?> v : JavaConversions.asJavaIterable(constraint.scope())) {
                v.domain_$eq(TrueDomain$.MODULE$);
                for (CSPOMConstraint c : JavaConversions.asJavaIterable(v.constraints())) {
                    if (c != constraint) {
                        constraints.add(c);
                    }
                }
            }
            problem.removeConstraint(constraint);

        }
    }

}

File: src/main/scala/cspom/compiler/patterns/RemoveAnd.scala
package cspom.compiler.patterns

import cspom.constraint.{ CSPOMConstraint, GeneralConstraint }
import cspom.variable.{ TrueDomain, CSPOMVariable }
import cspom.CSPOM
import scala.collection.mutable.Queue

class RemoveAnd(val problem: CSPOM, val constraints: Queue[CSPOMConstraint]) extends ConstraintCompiler {
  override def compile(constraint: CSPOMConstraint) {
    constraint match {
      case andConstraint: GeneralConstraint if constraint.description == "and" => {
        for (v <- constraint.scope) {
          v.asInstanceOf[CSPOMVariable].domain = TrueDomain
          for (c <- v.constraints if c != constraint)
            constraints.enqueue(c)
        }
        problem.removeConstraint(constraint);
      }
      case _ =>

    }
  }
}
